lib/common/model.js

```diff
@@ -553,6 +636,106 @@ function ModelFactory (
+                                var obj = {};
+                                obj[key] = {$each: [item]};
+                                obj.sign = sign;
+                                obj.field = key;
+                                (result.vaulesWithSign || (result.vaulesWithSign = []))
+                                .push(obj);
+                                (result.objsWithSign || (result.objsWithSign = []))
+                                .push(item);
+                            }
+                        });
+                    }, {});
+                }
+
+                result.vaulesWithSign = 
+                (result.vaulesWithSign || (result.vaulesWithSign = []))
+                .concat(withSign.vaulesWithSign || []);
+
+                (result.vaulesWithoutSign || (result.vaulesWithoutSign = {}))[key] = 
+                    {$each: require("lodash").difference(array, withSign.objsWithSign)};
+            }, {});
+
+            if (transformedValues.vaulesWithSign){
+                    transformedValues = transformedValues.vaulesWithSign
+                        .concat(transformedValues.vaulesWithoutSign);
+            }
+
+            return Promise.map(transformedValues, function(transformedValue){
+                var query = { _id: waterline.nodes.mongo.objectId(id) };
+                var update = {
+                    $addToSet: transformedValue
+                };
+                var options = {new: true};
+                if(transformedValue.sign){
+                    query.$not = 
+                        _.set({}, transformedValue.field, {$elemMatch: transformedValue.sign});
+                    transformedValue = _.pick(transformedValue, transformedValue.field);
+                    update = {
+                        $addToSet: transformedValue
+                    };
+                    return self.findAndModifyMongo(query, update, options);
+                }
+
+                return self.updateAndFindMongo(query, update, options);
+            }).then(function(modifiedDocs){
+                if (modifiedDocs.length === 1){
+                    return modifiedDocs[0];
+                } else {
+                    return self.needByIdentifier(id);
+                }
+            });
+        },
+
+        /**
+         * Using an id, find document and update fields only if they don't exist.
+         * Returns a waterline (bluebird) promise with the document find and modify.
+         *
+         * @param {string} id
+         * @param {String} field
+         * @param {Object/String/Number etc.} values
+         * @returns {Promise}
+         */
+        addFieldIfNotExistById: function (id, field, values) {
+            var self = this;
+
+            var query = { _id: waterline.nodes.mongo.objectId(id)};
+            query[field] = {$exists : false};
+
+            var update = { $set: {} };
+            update.$set[field] = values;
+
+            var options = {new: true};
```

Comment:
is "IfNotExist" necessary here? Can this function do the following things: 1) if the field doesn't exist, add it; 2) if it is already in the doc, return the original doc? In this case the callers don't need to check the field's existence by themselves in the upper layer.